private/buf/bufgraph/builder.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufgraph
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufanalysis"
+	"github.com/bufbuild/buf/private/bufpkg/bufimage"
+	"github.com/bufbuild/buf/private/bufpkg/bufimage/bufimagebuild"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
+	"github.com/bufbuild/buf/private/pkg/dag"
+	"go.uber.org/zap"
+)
+
+type builder struct {
+	logger         *zap.Logger
+	moduleResolver bufmodule.ModuleResolver
+	moduleReader   bufmodule.ModuleReader
+	imageBuilder   bufimagebuild.Builder
+}
+
+func newBuilder(
+	logger *zap.Logger,
+	moduleResolver bufmodule.ModuleResolver,
+	moduleReader bufmodule.ModuleReader,
+) *builder {
+	return &builder{
+		logger:         logger,
+		moduleResolver: moduleResolver,
+		moduleReader:   moduleReader,
+		imageBuilder: bufimagebuild.NewBuilder(
+			logger,
+			moduleReader,
+		),
+	}
+}
+
+func (b *builder) Build(
+	ctx context.Context,
+	modules []bufmodule.Module,
+	options ...BuildOption,
+) (*dag.Graph[Node], []bufanalysis.FileAnnotation, error) {
+	buildOptions := newBuildOptions()
+	for _, option := range options {
+		option(buildOptions)
+	}
+	return b.build(
+		ctx,
+		modules,
+		buildOptions.workspace,
+	)
+}
+
+func (b *builder) build(
+	ctx context.Context,
+	modules []bufmodule.Module,
+	workspace bufmodule.Workspace,
+) (*dag.Graph[Node], []bufanalysis.FileAnnotation, error) {
+	graph := dag.NewGraph[Node]()
+	for i, module := range modules {
+		fileAnnotations, err := b.buildForModule(
+			ctx,
+			module,
+			newNodeForModule(module, i),
+			workspace,
+			graph,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+		if len(fileAnnotations) > 0 {
+			return nil, fileAnnotations, nil
+		}
+	}
+	return graph, nil, nil
+}
+
+func (b *builder) buildForModule(
+	ctx context.Context,
+	module bufmodule.Module,
+	node Node,
+	workspace bufmodule.Workspace,
+	graph *dag.Graph[Node],
+) ([]bufanalysis.FileAnnotation, error) {
+	image, fileAnnotations, err := b.imageBuilder.Build(
+		ctx,
+		module,
+		bufimagebuild.WithWorkspace(workspace),
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(fileAnnotations) > 0 {
+		return fileAnnotations, nil
+	}
+	for _, imageModuleDependency := range bufimage.ImageModuleDependencies(image) {
```

Comment:
Just a question: Now that I look at it being used, would it be more natural to embed it into the `Image` interface?
```suggestion
	for _, imageModuleDependency := range image.ModuleDependencies() {
```

Reply:
It's half one way half the other - generally we haven't just added methods as we go to the interfaces, to try to keep the interfaces minimal - they only expose the minimal things you need to then be able to build up other concepts, such as ImageModuleDependencies. I'd like to keep it as-is for now as I can imagine this all getting rolled into `ImageFile` at some point (see comments).